Bad suggestion when `*raw_ptr` is used where a reference is expected

### Code

```Rust
fn foo(bar: *const [u8]) {
   let baz: &[u8] = unsafe { *bar };
}
```

### Current output

```Shell
error[E0308]: mismatched types
 --> src/lib.rs:2:30
  |
2 |    let baz: &[u8] = unsafe { *bar };
  |                              ^^^^ expected `&[u8]`, found `[u8]`
  |
help: consider removing deref here
  |
2 -    let baz: &[u8] = unsafe { *bar };
2 +    let baz: &[u8] = unsafe { bar };
  |
```

### Desired output

```Shell
error[E0308]: mismatched types
 --> src/lib.rs:2:30
  |
2 |    let baz: &[u8] = unsafe { *bar };
  |                              ^^^^ expected `&[u8]`, found `[u8]`
  |
help: consider creating a reference here
  |
2 -    let baz: &[u8] = unsafe { *bar };
2 +    let baz: &[u8] = unsafe { &*bar };
  |
```

### Rationale and extra context

The current suggestion is plain wrong, the result does not build.

Note that one definitely wants the `&` to go inside the unsafe block to avoid creating a temporary! For this example, the type is non-`Copy` so a temporary wouldn't even be possible, but in other cases both ways could compiler but they do not behave the same.

### Other cases

```Rust

```

### Rust Version

```Shell
current nightly
```

### Anything else?

_No response_

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"samueltardieu"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
@@ -2682,6 +2682,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
                     if let hir::ExprKind::Unary(hir::UnOp::Deref, inner) = expr.kind
                         && let Some(1) = self.deref_steps_for_suggestion(expected, checked_ty)
+                        && self.typeck_results.borrow().expr_ty(inner).is_ref()
                     {
                         // We have `*&T`, check if what was expected was `&T`.
                         // If so, we may want to suggest removing a `*`.

--- a/tests/ui/suggestions/raw-to-ref.fixed
+++ b/tests/ui/suggestions/raw-to-ref.fixed
@@ -1,0 +1,19 @@
+//@ run-rustfix
+// Regression test for #135580: check that we do not suggest to simply drop
+// the `*` to make the types match when the source is a raw pointer while
+// the target type is a reference.
+
+struct S;
+
+fn main() {
+    let mut s = S;
+    let x = &raw const s;
+    let _: &S = unsafe { &*x };
+    //~^ ERROR mismatched types
+    //~| HELP consider borrowing here
+
+    let x = &raw mut s;
+    let _: &mut S = unsafe { &mut *x };
+    //~^ ERROR mismatched types
+    //~| HELP consider mutably borrowing here
+}

--- a/tests/ui/suggestions/raw-to-ref.rs
+++ b/tests/ui/suggestions/raw-to-ref.rs
@@ -1,0 +1,19 @@
+//@ run-rustfix
+// Regression test for #135580: check that we do not suggest to simply drop
+// the `*` to make the types match when the source is a raw pointer while
+// the target type is a reference.
+
+struct S;
+
+fn main() {
+    let mut s = S;
+    let x = &raw const s;
+    let _: &S = unsafe { *x };
+    //~^ ERROR mismatched types
+    //~| HELP consider borrowing here
+
+    let x = &raw mut s;
+    let _: &mut S = unsafe { *x };
+    //~^ ERROR mismatched types
+    //~| HELP consider mutably borrowing here
+}

--- a/tests/ui/suggestions/raw-to-ref.stderr
+++ b/tests/ui/suggestions/raw-to-ref.stderr
@@ -1,0 +1,25 @@
+error[E0308]: mismatched types
+  --> $DIR/raw-to-ref.rs:11:26
+   |
+LL |     let _: &S = unsafe { *x };
+   |                          ^^ expected `&S`, found `S`
+   |
+help: consider borrowing here
+   |
+LL |     let _: &S = unsafe { &*x };
+   |                          +
+
+error[E0308]: mismatched types
+  --> $DIR/raw-to-ref.rs:16:30
+   |
+LL |     let _: &mut S = unsafe { *x };
+   |                              ^^ expected `&mut S`, found `S`
+   |
+help: consider mutably borrowing here
+   |
+LL |     let _: &mut S = unsafe { &mut *x };
+   |                              ++++
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0308`.

